sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala

```diff
@@ -117,14 +120,166 @@ private[hive] class HiveClientImpl(
+            override def run(): Hive = {
+              Hive.get(currentConf)
+            }
+          })
+        } else {
+          currentConf.setVar(ConfVars.METASTORE_EXECUTE_SET_UGI, user)
+          Hive.get(currentConf)
+        }
+        sessionHiveMap.put(user, hive)
+        hive
+      } finally {
+        Thread.currentThread().setContextClassLoader(original)
+      }
+    }
+  }
+
+  def obtainDelegationTokenRequired(hiveConf: HiveConf, ugi: UserGroupInformation): Boolean = {
+    UserGroupInformation.isSecurityEnabled &&
+      hiveConf.getTrimmed(ConfVars.METASTOREURIS.varname, "").nonEmpty &&
+      ugi.getAuthenticationMethod() == UserGroupInformation.AuthenticationMethod.PROXY &&
+      hiveConf.getBoolVar(HiveConf.ConfVars.METASTORE_USE_THRIFT_SASL)
+  }
+
+  /**
+   * Cached Hive Client initialized by login user
+   * This method only called under login UGI
+   *
+   * @return
+   */
+  private def cachedClient: Hive = {
+    if (clientLoader.cachedHive != null) {
+      clientLoader.cachedHive.asInstanceOf[Hive]
+    } else {
+      // here create hive client under login UGI and won't proxy user
+      val confWithoutImpersonation = state.getConf
+      confWithoutImpersonation.set("hive.metastore.token.signature", "")
+      val c = Hive.get(confWithoutImpersonation)
+      clientLoader.cachedHive = c
+      c
+    }
+  }
+
+  /**
+   * Get user's Delegation token for create new Hive Client with impersonation user
+   * Here we do this with [retryLocked] to avoid initial Hive client broken connection
+   * and retry connect
+   *
+   * @param user user need to get DelegationToken
+   * @return
+   */
+  private def getDelegationToken(user: String): String = clientLoader.synchronized {
+    // doAs under login user to make sure when hive client broken, it will reconnect
+    UserGroupInformation.getLoginUser.doAs[String](new PrivilegedExceptionAction[String] {
+      override def run(): String = {
+        // Hive sometimes retries internally, so set a deadline to avoid compounding delays.
+        val deadline = System.nanoTime + (retryLimit * retryDelayMillis * 1e6).toLong
+        var numTries = 0
+        var caughtException: Exception = null
+        do {
+          numTries += 1
+          try {
+            return cachedClient.getDelegationToken(user, user)
+          } catch {
+            case e: Exception if causedByThrift(e) =>
+              caughtException = e
+              logWarning(
+                "HiveClient got thrift exception, destroying client and retrying " +
+                  s"(${retryLimit - numTries} tries remaining)", e)
+              clientLoader.cachedHive = null
+              Thread.sleep(retryDelayMillis)
+          }
+        } while (numTries <= retryLimit && System.nanoTime < deadline)
+        if (System.nanoTime > deadline) {
+          logWarning("Deadline exceeded")
+        }
+        throw caughtException
+      }
+    })
+  }
+
```

Comment:
`cahedClient` is  initialed with login user, we use this client to get DelegationToken for other client calling user.
And do this with retry control.